### PR TITLE
Ported CnC-Campaign from our student project (now generalised)

### PR DESCRIPTION
**Pull request by Sithil-F**
Hello,
we are back with our campaign menu, but since we changed a lot of things, we decided to open a new pull request.
The original PR can be found here(https://github.com/OpenRA/OpenRA/pull/7559).
Here are the changes/additions from the last PR:
- Faction selection
  - the .vqa files are not mandatory anymore and we can remove them, if that is preferred
  - a custom recreation of the transmission-select screen if no vqa is found
- alternative mission selection (Campaign-Preview, i.e. a-,b-,c-missions)
  - the mission following a victory is declared in the yaml of the map (“Next Mission”), which allows for branching campaign-paths
  - If no preview-picture is available the faction logo appears
- Generalization, so it should work for all mods
  - currently active in the cnc and ra mods
- Other requests
  - use MapCache instead of manual preview
  - implement SaveProgressFlag
  - used OnPlayerWon Event to save the campaign progress

### CODE REVIEW

**Comment by Unit158** [comment]
You will need to split this up into manageable commits for the reviewers.

**Comment by pchote** [comment]
Adding new fields to Map and new game-level plumbing shouldn't be needed to support this.  We have created the mission.yaml files as a stub for holding this kind of info, and they should be expanded on instead.

**Comment by pchote** [comment]
The static CampaignProgress doesn't feel like a good fit.  See #7213 for some earlier discussion about recording our campaign progression.  I still think that the most flexible way to implement this will be via a Lua API for saving arbitrary data, which then lets us implement mission-objective specific progression (e.g. the C&C commando mission, which branched based on which structure you destroyed), and persisting information such as base layout across missions.

**Comment by pchote** [comment]
Can we please keep the windowed FMV playback (using buttons off the bottom of the campaign info panel) instead of reverting to the one-shot fullscreen videos?  We moved away from those because they do not look very good played at full size on modern high-resolution monitors, and because it is useful to be able to repeat the briefing if you miss something the first time.

**Comment by Sithil-F** [comment]
Thank you for your feedback. We will take care of the issues.
However, we will need help with the Lua API. 
(Could someone please mark this with the “PR: Help wanted” label?)
We will try to get the other 2 changes done within a week.

One Question about the outsourcing of the campaign paths:
So the possible paths should be defined in the missions.yaml, right?
Is the Country Information acceptable where it is, or should we move it too?

File: OpenRA.Game/Manifest.cs [hunk 1]
```diff
@@ -70,7 +70,7 @@ public class Manifest
 		readonly string[] reservedModuleNames = { "Metadata", "Folders", "MapFolders", "Packages", "Rules",
 			"Sequences", "VoxelSequences", "Cursors", "Chrome", "Assemblies", "ChromeLayout", "Weapons",
 			"Voices", "Notifications", "Music", "Translations", "TileSets", "ChromeMetrics", "Missions",
-			"ServerTraits", "LoadScreen", "LobbyDefaults", "Fonts", "TileSize", "MaximumTerrainHeight",
+			"FactionMedia",	"Congratulations",	"ServerTraits", "LoadScreen", "LobbyDefaults", "Fonts", "TileSize", "MaximumTerrainHeight",
```
Comment (abcdefg30) [inline OpenRA.Game/Manifest.cs]: Seems like you're using tabs instead of spaces after the commas.

File: OpenRA.Mods.Cnc/Widgets/Logic/CncMainMenuLogic.cs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#region Copyright & License Information
+﻿#region Copyright & License Information
```
Comment (abcdefg30) [inline OpenRA.Mods.Cnc/Widgets/Logic/CncMainMenuLogic.cs]: Can you undo the changes made to this file, please?

File: OpenRA.Mods.Common/CampaignProgress.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
+using System;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/CampaignProgress.cs]: The "Copyright & License Information" region is missing here.

File: OpenRA.Mods.Common/Widgets/Logic/CampaignCongratulationLogic.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Linq;
+
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/CampaignCongratulationLogic.cs]: Please remove this new line here.

File: OpenRA.Mods.Common/Widgets/Logic/CampaignFactionLogic.cs [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/CampaignFactionLogic.cs]: Here too.

**Comment by penev92** [comment]
>However, we will need help with the Lua API. 

What kind of help?

>So the possible paths should be defined in the missions.yaml, right?

Yes.

**Comment by penev92** [comment]
Bump.

**Comment by Freddiie** [comment]
Hello there,
we will begin in the coming days the implementation work on the Lua API. It would be helpful to know what kind of files and components to do with leaving a map in order to spare us familiarization time. Please have patience.

**Comment by penev92** [comment]
I'm going to close this now. Do reopen it if you find the time to finish it.

**Comment by obrakmann** [comment]
FWIW, I don't think this PR should be re-openend as-is. It's a massive pain to review such a huge thing. Especially the UI changes alone will take ages to get reviewed and be corrected. I suggest to pick and choose small parts of this and then pr them one-by-one.
